server/map_rpc_server.go

```diff
@@ -309,9 +308,21 @@ func (t *TrillianMapServer) getLeavesByRevision(ctx context.Context, mapID int64
 func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapLeavesRequest) (*trillian.SetMapLeavesResponse, error) {
 	ctx, spanEnd := spanFor(ctx, "SetLeaves")
 	defer spanEnd()
-
+	if req.Revision <= 0 {
+		return nil, status.Error(codes.FailedPrecondition, "revision must be > 0")
+	}
 	t.setLeafCounter.Add(float64(len(req.Leaves)), strconv.FormatInt(req.MapId, 10))
 
+	// TODO(pavelkalinnikov): Factor out a CreateRevision RPC in the MapWrite
```

Comment:
I don't agree with the TODO necessarily. In the commonly assumed case that each write batch will map to a published SMR revision then the sequencer/writer needs to make two calls to make a batch be published. When WriteLeaves becomes async, this adds the complexity of whether CreateRevision can be called for a revision which has not had its tree calculated yet.

I'm of the view that having a published revision per WriteLeaves batch massively simplifies the logical model of how the map works. The only downside to this is that WriteLeaves must be called with an empty batch to publish an SMR with a new timestamp but committing to the same contents. I don't think that's going to be very common, and it's still conceptually simple.

Reply:
> I don't agree with the TODO necessarily. In the commonly assumed case that each write batch will map to a published SMR revision then the sequencer/writer needs to make two calls to make a batch be published. When WriteLeaves becomes async, this adds the complexity of whether CreateRevision can be called for a revision which has not had its tree calculated yet.

I'm not proposing to call `CreateRevision` for every revision, only for empty ones. So it's still 1 RPC per revision.

Before:
1. `InitMap`
2. `WriteLeaves(rev=1, leaves=10)`, `WriteLeaves(rev=2, leaves=0)`, ....

After:
1. `CreateRevision(rev=0)`
2. `WriteLeaves(rev=1, leaves=10)`, `CreateRevision(rev=2)`, ...

 
> I'm of the view that having a published revision per WriteLeaves batch massively simplifies the logical model of how the map works. The only downside to this is that WriteLeaves must be called with an empty batch to publish an SMR with a new timestamp but committing to the same contents. I don't think that's going to be very common, and it's still conceptually simple.

The reason why someone would want to create a new SMR with 0 updates is refreshing the timestamp (I think that is the only one so far?). The reason for wanting to update the timestamp is that there haven't been any updates for quite a long time. In this case it makes sense to have a synchronous call like `CreateRevision(rev=N)` which immediately creates/returns an SMR.

The `WriteLeaves` call, if I understand the idea/design correctly, is not meant to be synchronous. It seems unreasonable allowing to create an empty SMR "at some point in the future": it's either needed "right now", or not needed at all (in case `WriteLeaves` was called in the meantime, and the new SMR construction is ongoing anyway).

Reply:
The purpose of this TODO was to express my "meh" to the API that we have right now. I made it a "Consider" type of TODO, because there are other options, which include decoupling SMRs from revisions (I don't see a strong reason why signing should be part of updating the tree), or allowing multiple SMRs per revision. Since the next gen map is in the design phase, all options should be on the table.

Moved this TODO to `addRevision` method comment.